File: java/src/main/java/util/LimitUtil.java
package util;

import java.util.Arrays;
import java.util.List;

public class LimitUtil {
    private static volatile int seq = 0;

    private static final int bucket = 1000;

    private static volatile List<Long> requestTimeList = Arrays.asList(0L, 0L, 0L, 0L, 0L);

    private LimitUtil() {
    }

    /**
     * 限流实现 用 static synchronized 修饰 表示 对类上锁
     *
     * @throws InterruptedException
     */
    public static synchronized void tryBeforeRun() throws InterruptedException {
        long now = System.currentTimeMillis();
        // 当前时间 与 前limit个请求的时间 作比较
        long interval = now - requestTimeList.get(seq);
        if (interval < 0) {
            // 执行时间: t[seq]+bucket=now-interval+bucket
            // 等待时间: bucket-interval
            Thread.sleep(bucket - interval);
            // 重新等待
            tryBeforeRun();
        }
        if (interval < bucket) {
            requestTimeList.set(seq, requestTimeList.get(seq) + bucket);
            Thread.sleep(bucket - interval);
        } else {
            requestTimeList.set(seq, now);
        }
        seq = (seq + 1) % requestTimeList.size();
    }
}

File: python/src/util/limit_util.py
import time


# 限流器
class Limiter:
    def __init__(self, duration, limit):
        self.seq = 0
        self.bucket = duration
        self.request_time_list = [0 for x in range(0, limit)]

    def tryBeforeRun(self):
        now_mill_time = int(round(time.time() * 1000))
        interval = now_mill_time - self.request_time_list[self.seq]
        if interval < 0:
            #  这边是毫秒数 需要修改成秒
            time.sleep((self.bucket - interval) / 1000)
            self.tryBeforeRun()
        if interval < self.bucket:
            self.request_time_list[self.seq] = self.request_time_list[self.seq] + self.bucket
            #  这边是毫秒数 需要修改
            time.sleep((self.bucket - interval) / 1000)
        else:
            self.request_time_list[self.seq] = now_mill_time
        self.seq = (self.seq + 1) % len(self.request_time_list)
